src/sign.ml

```diff
@@ -264,18 +269,36 @@ let read : string -> t = fun fname ->
 (* NOTE here, we rely on the fact that a marshaled closure can only be read by
    processes running the same binary as the one that produced it. *)
 
-(** [add_rule sign sym r] adds the new rule [r] to the symbol [sym].  When the
-    rule does not correspond to a symbol of signature [sign],  it is stored in
-    its dependencies. *)
-let add_rule : t -> sym -> rule -> unit = fun sign sym r ->
-  sym.sym_rules := !(sym.sym_rules) @ [r];
-  if sym.sym_path <> sign.sign_path then
-    let m =
-      try PathMap.find sym.sym_path !(sign.sign_deps)
-      with Not_found -> assert false
-    in
-    let m = (sym.sym_name, r) :: m in
-    sign.sign_deps := PathMap.add sym.sym_path m !(sign.sign_deps)
+(** [add_rules s r] adds a bunch of rules [r] to its attached symbol to the
+    signature [s] and builds decision trees. *)
+let add_rules : t -> (sym * pp_hint * rule loc) list -> unit = fun sign rs ->
+  (* [add_rule sign sym r] adds the new rule [r] to the symbol [sym].  When
+     the rule does not correspond to a symbol of signature [sign], it is
+     stored in its dependencies. *)
+  let add_rule (s, h, r) =
+    out 3 "(rule) %a\n" Print.pp_rule (s,h,r.elt) ;
+    s.sym_rules := !(s.sym_rules) @ [r.elt] ;
+    if s.sym_path <> sign.sign_path then
+      let m =
+        try PathMap.find s.sym_path !(sign.sign_deps)
+        with Not_found -> assert false in
+      let m = (s.sym_name, r.elt) :: m in
+      sign.sign_deps := PathMap.add s.sym_path m !(sign.sign_deps) in
+  List.iter add_rule rs ;
+  let build_tree symb =
+    match symb.sym_mode with
+    | Defin
+    | Injec ->
+        let pama = lazy (Dtree.ClauseMat.of_rules !(symb.sym_rules)) in
+        let tree = lazy (Dtree.compile @@ Lazy.force pama) in
+        let capacity = lazy (Basics.capacity @@ Lazy.force tree) in
+        symb.sym_tree := (capacity, tree) ;
+        if Pervasives.(!write_trees) then
+          ( Format.printf "Wrote %s.gv\n" (symb.sym_name)
+          ; Dtree.to_dot symb.sym_name (Lazy.force tree) )
+    | _     -> () in
+  List.map (fun (e, _, _) -> e) rs |> List.sort_uniq Basics.sym_cmp |>
```

Comment:
Can be easily improved by removing List.map and locally defining a comparison function on triples. Another further amelioration is to compute a set of symbols while adding rules and then iterate on this set.

Reply:
The first suggestion has been performed